src/nupic/utils/GroupBy.hpp

```diff
@@ -0,0 +1,1241 @@
+   * containing the key, followed by a begin and end iterator for each
+   * sequence. The sequences are traversed lazily as the iterator is advanced.
+   *
+   * Feel free to add new GroupBy7, GroupBy8, ..., GroupByN classes as needed.
+   */
+
+  // ==========================================================================
+  // 1 SEQUENCE
+  // ==========================================================================
+
+ template<typename Iterator0, typename KeyFn0,
+           typename Element0 = decltype(*std::declval<Iterator0>()),
+           typename KeyType = typename std::result_of<
+             KeyFn0(decltype(*std::declval<Iterator0>()))
+             >::type>
+  class GroupBy1
+  {
+  public:
+
+    GroupBy1(
+      Iterator0 begin0, Iterator0 end0, KeyFn0 keyFn0)
+    : begin0_(begin0), end0_(end0), keyFn0_(keyFn0)
+    {
+      NTA_ASSERT(std::is_sorted(begin0, end0,
+                                [&](const Element0& a, const Element0& b)
+                                {
+                                  return keyFn0(a) < keyFn0(b);
+                                }));
+    }
+
+    class Iterator
+    {
+    public:
+      Iterator(
+        Iterator0 begin0, Iterator0 end0, KeyFn0 keyFn0)
+        :current0_(begin0), end0_(end0), keyFn0_(keyFn0),
+         finished_(false)
+      {
+        calculateNext_();
+      }
+
+      bool operator !=(const Iterator& other)
+      {
+        return (finished_ != other.finished_ ||
+                current0_ != other.current0_);
+      }
+
+      const std::tuple<KeyType,
+                       Iterator0, Iterator0>& operator*() const
+      {
+        NTA_ASSERT(!finished_);
+        return v_;
+      }
+
+      const Iterator& operator++()
+      {
+        NTA_ASSERT(!finished_);
+        calculateNext_();
+        return *this;
+      }
+
+    private:
+
+      void calculateNext_()
+      {
+        if (current0_ != end0_)
+        {
+          KeyType key = keyFn0_(*current0_);
+          std::get<0>(v_) = key;
+
+          // Find all elements with this key.
+          std::get<1>(v_) = current0_;
+          while (current0_ != end0_ && keyFn0_(*current0_) == key)
+          {
+            current0_++;
+          }
+          std::get<2>(v_) = current0_;
+        }
+        else
+        {
+          finished_ = true;
+        }
+      }
+
+      std::tuple<KeyType,
+                 Iterator0, Iterator0> v_;
+
+      Iterator0 current0_;
+      Iterator0 end0_;
+      KeyFn0 keyFn0_;
+
+      bool finished_;
+    };
+
+    Iterator begin() const
+    {
+      return Iterator(begin0_, end0_, keyFn0_);
+    }
+
+    Iterator end() const
+    {
+      return Iterator(end0_, end0_, keyFn0_);
+    }
+
+  private:
+    Iterator0 begin0_;
+    Iterator0 end0_;
+    KeyFn0 keyFn0_;
+  };
+
+  template<typename Sequence0, typename KeyFn0>
+  GroupBy1<typename Sequence0::const_iterator, KeyFn0>
+  group_by(const Sequence0& sequence0, KeyFn0 keyFn0)
```

Comment:
I added a note to GroupBy.hpp:

~~~c++
  /*
   * Note: It's called "group_by" and "iter_group_by" in snake_case because it's
   * designed to feel like an STL function, like "is_sorted" or "iter_swap".
   */
~~~